sensor/common/collector/service_impl.go

```diff
@@ -0,0 +1,141 @@
+	"github.com/stackrox/rox/sensor/common"
+	"github.com/stackrox/rox/sensor/common/message"
+	"google.golang.org/grpc"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+// CollectorService is the struct that manages the collector configuration
+type serviceImpl struct {
+	sensor.UnimplementedCollectorServiceServer
+
+	collectorC chan *sensor.MsgToCollector
+
+	connectionManager *connectionManager
+}
+
+func (s *serviceImpl) Notify(e common.SensorComponentEvent) {
+}
+
+func (s *serviceImpl) Start() error {
+	s.collectorC = make(chan *sensor.MsgToCollector)
+	return nil
+}
+
+func (s *serviceImpl) Stop(_ error) {
+	close(s.collectorC)
+}
+
+func (s *serviceImpl) Capabilities() []centralsensor.SensorCapability {
+	return nil
+}
+
+func getCollectorConfig(msg *central.MsgToSensor) *storage.CollectorConfig {
+	if clusterConfig := msg.GetClusterConfig(); clusterConfig != nil {
+		if config := clusterConfig.GetConfig(); config != nil {
+			if collectorConfig := config.GetCollectorConfig(); collectorConfig != nil {
+				return collectorConfig
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *serviceImpl) ProcessMessage(msg *central.MsgToSensor) error {
+	log.Info("In ProcessMessage")
+	if collectorConfig := getCollectorConfig(msg); collectorConfig != nil {
+		log.Infof("Sending message %+v ", collectorConfig)
+		s.collectorC <- &sensor.MsgToCollector{
+			Msg: &sensor.MsgToCollector_CollectorConfig{
+				CollectorConfig: collectorConfig,
+			},
+		}
+	}
+
+	return nil
+}
+
+func (s *serviceImpl) ResponsesC() <-chan *message.ExpiringMessage {
+	return nil
+}
+
+type connectionManager struct {
+	connectionLock sync.RWMutex
+	connectionMap  map[sensor.CollectorService_CommunicateServer]bool
+}
+
+func newConnectionManager() *connectionManager {
+	return &connectionManager{
+		connectionMap: make(map[sensor.CollectorService_CommunicateServer]bool),
+	}
+}
+
+func (c *connectionManager) add(connection sensor.CollectorService_CommunicateServer) {
+	c.connectionLock.Lock()
+	log.Info("Adding connection")
+	log.Infof("connection= %+v", connection)
+	defer c.connectionLock.Unlock()
+
+	c.connectionMap[connection] = true
+}
+
+func (c *connectionManager) remove(connection sensor.CollectorService_CommunicateServer) {
+	c.connectionLock.Lock()
+	log.Info("Removing connection")
+	log.Infof("connection= %+v", connection)
+	defer c.connectionLock.Unlock()
+
+	delete(c.connectionMap, connection)
+}
+
+func (s *serviceImpl) Communicate(server sensor.CollectorService_CommunicateServer) error {
+
+	s.connectionManager.add(server)
+	defer s.connectionManager.remove(server)
+	log.Info("In Communicate")
+
+	for msg := range s.collectorC {
+		log.Info("Sending message")
+		log.Infof("len(s.connectionManager.connectionMap)= %+v", len(s.connectionManager.connectionMap))
+		for conn := range s.connectionManager.connectionMap {
+			err := conn.Send(msg)
+			if err != nil {
+				log.Error(err, "Failed sending runtime config to Collector")
+				return err
+			}
+		}
+	}
```

Comment:
This will block the goroutine and the Communicate function will be only able to send, but will not be receiving anything back from Collector. Is that the intention, here?

Note that the current implementation would make it difficult to implement any protocol where a given order of request-reply must be preserved. This may be fine, but may also be a problem - that depend on the future plans.

I know that we have a separate connection from Collector to Sensor, but then something feels weird here...
We implement here a `service` and services are mainly for handling the calls originating in the other communicating party. However the core job of this service is to send messages to Collector. I feel like I still need to think more about that part, but maybe we could try to get the inspiration not from Sensor-Compliance communication, but instead from the "hello" protocol that we use between Sensor and Central? It would provide a GRPC stream and a way to exchange information in the initial handshake. That would give us the guarantee that the inital handshake and config exchange would always happen before the regular stream would start. As said, this still requires some thinking...

Reply:
This is intentional. There is no communication from collector planned. If that changes in the future we can make this non-blocking by adding `go`.